designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,956 @@
+1. Merging of config files via `extends`
+1. Overriding of configuration via `overrides`
+
+All of these complexities arise from `.eslintrc` format of describing what should happen rather than how it should happen. We keep trying to anticipate how users want things to happen and guessing wrong has led to increased complexity.
+
+This RFC proposes a fresh start for configuration in ESLint that takes into account all of the requests we've received over the years. By simplifying and stripping configuration we can actually make a configuration system that is flexible enough to accomodate changes in the future and easy enough to implement that we won't be afraid to do so.
+
+## Detailed Design
+
+Design Summary:
+
+1. Introduce a new `eslint.config.js` configuration file format
+1. Searching for `eslint.config.js` starts from the current working directory and continues up
+1. All `eslint.config.js` files are treated as if they have `root: true`
+1. There is no automatic merging of config files
+1. The `eslint.config.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
+1. An `@eslint/eslintrc` utility is provided to aid with backwards compatibility
+1. Remove the concept of environments (`env`) (including `--env`, `/*eslint-env*/`, and `env` in config files)
+1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
+1. Remove `--rulesdir`
+1. Remove `--ext`
+1. Remove `--resolve-plugins-relative-to`
+1. Replace `context.parserPath` with `context.parser` (the path to the parser is no longer valid)
+1. Allow plugins to specify a `parsers` key so parsers can be referenced the same as rules and processors in plugins.
+
+### The `eslint.config.js` File
+
+The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports an object: 
+
+```js
+module.exports = {
+    name: "name",
+    files: ["*.js"],
+    ignores: ["*.test.js"],
+    settings: {},
+    languageOptions: {
+        ecmaVersion: 2020,
+        sourceType: "module",
+        globals: {},
+        parser: object || "string",
+        parserOptions: {},
+        linterOptions: {
+            reportUnusedDisableDirectives: "string"
+        }
+    }
+    processor: object || "string",
+    plugins: {}
+    rules: {}
+};
+```
+
+The following keys are new to the `eslint.config.js` format:
+
+* `name` - Specifies the name of the config object. This is helpful for printing out debugging information and, while not required, is recommended for that reason.
+* `files` - Determines the glob file patterns that this configuration applies to.
+* `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.
```

Comment:
`ignores` acts like `excludedFiles` when `files` is present.

Reply:
Thanks. So does the `ignores` property have different meanings between the `files` property is present and not?

- If the `files` is present, ESLint uses the `ignores` as `excludedFiles` before it merges config array elements.
- Otherwise, ESLint merges the `ignores` as well, then it doesn't lint the files that match the merged `ignores` at all.

If so, I don't think it's good behavior.

Reply:
It's really the same meaning:

* if `files` is present: match patterns in `files` and then exclude things in `ignores`.
* if `files` is not present: assume `files` is "*" and then exclude things in `ignores`.

So the matching works the same way in both cases, we just assume you meant for `files` to apply to everything if it's not present.



As a reminder: the prototype already took this into account and things worked fine.

Reply:
No. I quote this line.

> Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.

1. The "Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file." part describes the functionality of `.eslintignore` and the `ignorePatterns` in eslintrc.
1. The "The files specified by this array of glob patterns are subtracted from the files specified in `files`." part describes the functionality of the `excludedFiles` field of eslintrc.
1. The "If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files." part describes the functionality of `.eslintignore` and the `ignorePatterns` in eslintrc again.

I feel this is really confusing.

The `excludedFiles` and `ignorePatterns` (`.eslintignore`) fields in eslintrc are pretty different.

```js
module.exports = {
    {
        files: ["*.js"],
        rules: { /* rules for all code */ }
    },
    {
        // files: ["*.js"],
        ignores: ["*.spec.js"],
        rules: { /* rules for production js code */ }
    },
}
```

With the above config, the second entry doesn't have the `files` field, so I assume it works as same as `.eslintignore`. Therefore, ESLint doesn't lint `*.spec.js` at all. Because the purpose of `.eslintignore` is to specify ignored files/dirs such as `node_modules`, `dist`, or `build_cache`.

On the other hand, if I assume the `ignores` field behaves as same as the `excludedFiles` field of eslintrc, with the above config, ESLint lints `*.spec.js` files with using only the first entry. ESLint doesn't ignore `*.spec.js` files.

----

> - if files is present: match patterns in files and then exclude things in ignores.
> - if files is not present: assume files is "*" and then exclude things in ignores.

It's the description of `excludedFiles`. It's pretty different from `.eslintignore`/`ignorePatterns`.

Reply:
> As a reminder: the prototype already took this into account and things worked fine.

The prototype doesn't work for me.

```
> npm i eslint/eslint#simpleconfig
> eslint test.js
Error: Key "ignores" requires keys "files".
    at ObjectSchema.validate (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\@humanwhocodes\object-schema\src\object-schema.js:157:27)
    at C:\Users\t-nagashima.AD\dev\sandbox\node_modules\@humanwhocodes\object-schema\src\object-schema.js:116:18
    at Array.reduce (<anonymous>)
    at ObjectSchema.merge (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\@humanwhocodes\object-schema\src\object-schema.js:114:24)
    at C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\simpleconfig\config-array.js:114:27
    at Array.reduce (<anonymous>)
    at ConfigArray.getConfig (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\simpleconfig\config-array.js:113:32)
    at processText (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\cli-engine.js:183:33)
    at processFile (C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\cli-engine.js:246:12)
    at C:\Users\t-nagashima.AD\dev\sandbox\node_modules\eslint\lib\cli-engine.js:634:40
```

Reply:
> On the other hand, if I assume the ignores field behaves as same as the excludedFiles field of eslintrc, with the above config, ESLint lints *.spec.js files with using only the first entry. ESLint doesn't ignore *.spec.js files.

Just to make sure I understand: you're saying that `excludedFiles` does not require the use of `files` for overrides? In that case, the override will always apply *except* for the files in `excludedFiles`?

Reply:
> Just to make sure I understand: you're saying that excludedFiles does not require the use of files for overrides? In that case, the override will always apply except for the files in excludedFiles?

No. We cannot omit the `files` field in `overrides` of the current config system. I meant that, if I interpret that the `ignores` corresponds to the current `excludedFiles`, ...

In the current config system, `ignorePatterns` and `excludedFiles` are different stuff:

- `ignorePatterns`:
  - A list of gitignore-style patterns. (therefore, we can read .gitignore and set it.)
  - All of the `ignorePatterns` field are merged. It's used *after* all config array elements are merged to one.
  - It supports negative patterns to unignore other ignore patterns. For example, `!foo.js` will unignore `foo.js` even if a shareable config is configured to ignore it.
- `excludedFiles`:
  - A list of glob patterns.
  - Every `excludedFiles` field is used to check the adoption of each `overrides` entry. It's used *before* all config array elements are merged to one.
  - It doesn't support negative patterns.

This RFC says a mix of those for the `ignores` field. Therefore, I'm confusing.

- If `files` field exists, the `ignores` field is used for the adoption of the config array element and isn't merged for the check of ignored files. It's glob patterns as the RFC says.
- If `files` field doesn't exist, the `ignores` field doesn't affect the adoption of the config array element and is merged for the check ignored files. It's a list of gitignore-style patterns and supports negative patterns, as same as `.eslintignore`/`ignorePatterns`.